File: 5-saga-orchestration-and-choreography/inventory-service/src/stock/stock.controller.ts
/**
 * HTTP controller — debug routes delegate to service.
 * (EN: HTTP controller — debug routes delegating to service.)
 */
import {
    Body,
    Controller,
    Get,
    Post,
} from "@nestjs/common"
import {
    InjectRepository,
} from "@nestjs/typeorm"
import {
    Repository,
} from "typeorm"
import {
    ProductEntity,
} from "../entities"
import {
    StockService,
} from "./stock.service"

@Controller()
export class StockController {
    constructor(
        private readonly stock: StockService,
        @InjectRepository(ProductEntity) private readonly products: Repository<ProductEntity>,
    ) {}

    /**
     * Logic — đọc tồn kho hiện tại (debug).
     * Code — GET /stock → products.find().
     * (EN Logic: Read current stock levels (debug).)
     * (EN Code: GET /stock → products.find().)
     */
    @Get("stock")
    async list(): Promise<ProductEntity[]> {
        return this.products.find()
    }

    /**
     * Logic — gọi thử bước inventory của saga (debug).
     * Code — POST /check → stock.tryFulfill().
     * (EN Logic: Manually trigger the inventory saga step (debug).)
     * (EN Code: POST /check → stock.tryFulfill().)
     */
    @Post("check")
    async check(
        @Body() body: { orderId: number; productId: number; quantity: number },
    ): ReturnType<StockService["tryFulfill"]> {
        return this.stock.tryFulfill(body.orderId, body.productId, body.quantity)
    }
}

File: 5-saga-orchestration-and-choreography/1-java/inventory-service/src/main/java/com/starci/inventory/StockController.java
package com.starci.inventory;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

/**
 * StockController — REST controller for manual inventory check.
 *
 * <p>Mirrors the TS lesson's {@code POST /check} endpoint which lets callers
 * manually trigger a stock reservation outside of the Kafka flow.
 */
@RestController
public class StockController {

    private final StockService stock;

    /**
     * Constructor injection.
     *
     * @param stock service containing saga inventory logic
     */
    public StockController(StockService stock) {
        this.stock = stock;
    }

    /**
     * Manually trigger a stock reservation check.
     *
     * <p>Logic: delegate to StockService.tryFulfill() which handles idempotency,
     * validation, MongoDB update, and Kafka event emission.
     * Code: POST /check → StockService.tryFulfill() → return result.
     *
     * @param body request body with orderId, productId, and quantity
     * @return fulfillment result
     */
    @PostMapping("/check")
    public ResponseEntity<InventoryCheckResult> check(@RequestBody CheckRequest body) {
        InventoryCheckResult result = stock.tryFulfill(
                body.getOrderId(), body.getProductId(), body.getQuantity());
        return ResponseEntity.ok(result);
    }
}
